File: examples/standalone/go/10-path-router.go
package main

import "strings"

// Example 10: HTTP path routing and middleware resolution.
// Tests reasoning about string patterns, slice lookups, and multi-condition dispatch.

// Route defines a route pattern with method and handler name.
type Route struct {
	Method  string
	Pattern string
	Handler string
}

// MatchResult holds the outcome of a route match attempt.
type MatchResult struct {
	Status  string
	Handler string
	Params  map[string]string
	Message string
}

// MatchRoute — 12 branches: empty path→error, no leading /→error,
// empty method→error, no routes→not-found, exact static match→matched,
// param match (:id)→matched+params, wildcard (*)→matched+wildcard,
// method mismatch→method-not-allowed, no match→not-found,
// first match wins, trailing slash normalized, multiple params extracted.
func MatchRoute(routes []Route, method string, path string) MatchResult {
	if len(path) == 0 {
		return MatchResult{Status: "error", Message: "empty path"}
	}
	if !strings.HasPrefix(path, "/") {
		return MatchResult{Status: "error", Message: "path must start with /"}
	}
	if len(method) == 0 {
		return MatchResult{Status: "error", Message: "empty method"}
	}
	if len(routes) == 0 {
		return MatchResult{Status: "not-found"}
	}

	normalizedPath := path
	if len(path) > 1 && strings.HasSuffix(path, "/") {
		normalizedPath = path[:len(path)-1]
	}

	pathSegments := splitSegments(normalizedPath)
	upperMethod := strings.ToUpper(method)
	pathMatchedButMethodDiffers := false

	for _, route := range routes {
		patternSegments := splitSegments(route.Pattern)

		if len(patternSegments) > 0 && patternSegments[len(patternSegments)-1] == "*" {
			prefixSegments := patternSegments[:len(patternSegments)-1]
			if len(pathSegments) >= len(prefixSegments) {
				prefixMatches := true
				params := make(map[string]string)

				for i := 0; i < len(prefixSegments); i++ {
					ps := prefixSegments[i]
					if strings.HasPrefix(ps, ":") {
						params[ps[1:]] = pathSegments[i]
					} else if ps != pathSegments[i] {
						prefixMatches = false
						break
					}
				}

				if prefixMatches {
					if strings.ToUpper(route.Method) == upperMethod {
						params["wildcard"] = strings.Join(pathSegments[len(prefixSegments):], "/")
						return MatchResult{Status: "matched", Handler: route.Handler, Params: params}
					}
					pathMatchedButMethodDiffers = true
				}
			}
			continue
		}

		if len(patternSegments) != len(pathSegments) {
			continue
		}

		matches := true
		params := make(map[string]string)
		for i := 0; i < len(patternSegments); i++ {
			ps := patternSegments[i]
			if strings.HasPrefix(ps, ":") {
				params[ps[1:]] = pathSegments[i]
			} else if ps != pathSegments[i] {
				matches = false
				break
			}
		}

		if matches {
			if strings.ToUpper(route.Method) == upperMethod {
				return MatchResult{Status: "matched", Handler: route.Handler, Params: params}
			}
			pathMatchedButMethodDiffers = true
		}
	}

	if pathMatchedButMethodDiffers {
		return MatchResult{Status: "method-not-allowed"}
	}
	return MatchResult{Status: "not-found"}
}

func splitSegments(path string) []string {
	var segments []string
	for _, s := range strings.Split(path, "/") {
		if len(s) > 0 {
			segments = append(segments, s)
		}
	}
	return segments
}

// RouteMetadata describes middleware requirements for a route.
type RouteMetadata struct {
	RequiresAuth   bool
	ContentType    string
	CorsOrigin     string
	HasCorsOrigin  bool
	AllowedOrigins []string
}

// MiddlewareChain holds the resolved middleware list.
type MiddlewareChain struct {
	Status     string
	Middleware []string
	Reason     string
}

// ResolveMiddleware — 8 branches: auth required+no header→rejected,
// auth+invalid format→rejected, auth+valid→adds auth middleware,
// json content→adds json-parser, multipart→adds multipart-parser,
// cors+allowed origin→adds cors, cors+disallowed→rejected, base chain only.
func ResolveMiddleware(metadata RouteMetadata, authHeader string) MiddlewareChain {
	chain := []string{"logging", "error-handler"}

	if metadata.RequiresAuth {
		if len(authHeader) == 0 {
			return MiddlewareChain{Status: "rejected", Middleware: nil, Reason: "missing auth"}
		}
		if !strings.HasPrefix(authHeader, "Bearer ") || len(authHeader) <= 7 {
			return MiddlewareChain{Status: "rejected", Middleware: nil, Reason: "invalid auth format"}
		}
		chain = append(chain, "auth")
	}

	if metadata.ContentType == "application/json" {
		chain = append(chain, "json-parser")
	} else if metadata.ContentType == "multipart/form-data" {
		chain = append(chain, "multipart-parser")
	}

	if metadata.HasCorsOrigin {
		found := false
		for _, origin := range metadata.AllowedOrigins {
			if origin == metadata.CorsOrigin {
				found = true
				break
			}
		}
		if found {
			chain = append(chain, "cors")
		} else {
			return MiddlewareChain{Status: "rejected", Middleware: nil, Reason: "origin not allowed"}
		}
	}

	return MiddlewareChain{Status: "ok", Middleware: chain}
}

File: examples/standalone/rust/10_path_router.rs
// Example 10: HTTP path routing and middleware resolution.
// Tests reasoning about string patterns, vector lookups, and multi-condition dispatch.

use std::collections::HashMap;

struct Route {
    method: String,
    pattern: String,
    handler: String,
}

struct MatchResult {
    status: &'static str,
    handler: Option<String>,
    params: HashMap<String, String>,
    message: Option<String>,
}

/// match_route — 12 branches: empty path→error, no leading /→error,
/// empty method→error, no routes→not-found, exact static→matched,
/// param (:id)→matched+params, wildcard (*)→matched+wildcard,
/// method mismatch→method-not-allowed, no match→not-found,
/// first match wins, trailing slash normalized, multiple params extracted.
fn match_route(routes: &[Route], method: &str, path: &str) -> MatchResult {
    if path.is_empty() {
        return MatchResult {
            status: "error",
            handler: None,
            params: HashMap::new(),
            message: Some("empty path".to_string()),
        };
    }
    if !path.starts_with('/') {
        return MatchResult {
            status: "error",
            handler: None,
            params: HashMap::new(),
            message: Some("path must start with /".to_string()),
        };
    }
    if method.is_empty() {
        return MatchResult {
            status: "error",
            handler: None,
            params: HashMap::new(),
            message: Some("empty method".to_string()),
        };
    }
    if routes.is_empty() {
        return MatchResult {
            status: "not-found",
            handler: None,
            params: HashMap::new(),
            message: None,
        };
    }

    let normalized = if path.len() > 1 && path.ends_with('/') {
        &path[..path.len() - 1]
    } else {
        path
    };

    let path_segments = split_segments(normalized);
    let upper_method = method.to_uppercase();
    let mut path_matched_but_method_differs = false;

    for route in routes {
        let pattern_segments = split_segments(&route.pattern);

        if !pattern_segments.is_empty() && pattern_segments.last() == Some(&"*".to_string()) {
            let prefix_segments = &pattern_segments[..pattern_segments.len() - 1];
            if path_segments.len() >= prefix_segments.len() {
                let mut prefix_matches = true;
                let mut params = HashMap::new();

                for (i, ps) in prefix_segments.iter().enumerate() {
                    if ps.starts_with(':') {
                        params.insert(ps[1..].to_string(), path_segments[i].clone());
                    } else if *ps != path_segments[i] {
                        prefix_matches = false;
                        break;
                    }
                }

                if prefix_matches {
                    if route.method.to_uppercase() == upper_method {
                        params.insert(
                            "wildcard".to_string(),
                            path_segments[prefix_segments.len()..].join("/"),
                        );
                        return MatchResult {
                            status: "matched",
                            handler: Some(route.handler.clone()),
                            params,
                            message: None,
                        };
                    }
                    path_matched_but_method_differs = true;
                }
            }
            continue;
        }

        if pattern_segments.len() != path_segments.len() {
            continue;
        }

        let mut matches = true;
        let mut params = HashMap::new();
        for (i, ps) in pattern_segments.iter().enumerate() {
            if ps.starts_with(':') {
                params.insert(ps[1..].to_string(), path_segments[i].clone());
            } else if *ps != path_segments[i] {
                matches = false;
                break;
            }
        }

        if matches {
            if route.method.to_uppercase() == upper_method {
                return MatchResult {
                    status: "matched",
                    handler: Some(route.handler.clone()),
                    params,
                    message: None,
                };
            }
            path_matched_but_method_differs = true;
        }
    }

    if path_matched_but_method_differs {
        return MatchResult {
            status: "method-not-allowed",
            handler: None,
            params: HashMap::new(),
            message: None,
        };
    }

    MatchResult {
        status: "not-found",
        handler: None,
        params: HashMap::new(),
        message: None,
    }
}

fn split_segments(path: &str) -> Vec<String> {
    path.split('/')
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .collect()
}

struct RouteMetadata {
    requires_auth: bool,
    content_type: String,
    cors_origin: Option<String>,
    allowed_origins: Vec<String>,
}

struct MiddlewareChain {
    status: &'static str,
    middleware: Vec<String>,
    reason: Option<String>,
}

/// resolve_middleware — 8 branches: auth+no header→rejected,
/// auth+invalid format→rejected, auth+valid→adds "auth",
/// json→adds "json-parser", multipart→adds "multipart-parser",
/// cors+allowed→adds "cors", cors+disallowed→rejected, base chain only.
fn resolve_middleware(metadata: &RouteMetadata, auth_header: Option<&str>) -> MiddlewareChain {
    let mut chain = vec!["logging".to_string(), "error-handler".to_string()];

    if metadata.requires_auth {
        match auth_header {
            None | Some("") => {
                return MiddlewareChain {
                    status: "rejected",
                    middleware: vec![],
                    reason: Some("missing auth".to_string()),
                };
            }
            Some(h) => {
                if !h.starts_with("Bearer ") || h.len() <= 7 {
                    return MiddlewareChain {
                        status: "rejected",
                        middleware: vec![],
                        reason: Some("invalid auth format".to_string()),
                    };
                }
                chain.push("auth".to_string());
            }
        }
    }

    if metadata.content_type == "application/json" {
        chain.push("json-parser".to_string());
    } else if metadata.content_type == "multipart/form-data" {
        chain.push("multipart-parser".to_string());
    }

    if let Some(origin) = &metadata.cors_origin {
        if metadata.allowed_origins.iter().any(|o| o == origin) {
            chain.push("cors".to_string());
        } else {
            return MiddlewareChain {
                status: "rejected",
                middleware: vec![],
                reason: Some("origin not allowed".to_string()),
            };
        }
    }

    MiddlewareChain {
        status: "ok",
        middleware: chain,
        reason: None,
    }
}

fn main() {
    let routes = vec![
        Route {
            method: "GET".to_string(),
            pattern: "/users/:id".to_string(),
            handler: "get_user".to_string(),
        },
    ];
    let result = match_route(&routes, "GET", "/users/42");
    println!("status={}, handler={:?}, params={:?}", result.status, result.handler, result.params);
}
